tst/unit/test_state_descriptor.cpp

```diff
@@ -0,0 +1,270 @@
+          REQUIRE(pkg3->SparsePresent("package1::sparse"));
+          REQUIRE(pkg3->SparsePresent("package2::sparse"));
+          REQUIRE(!(pkg3->SparsePresent("sparse")));
+        }
+        AND_THEN("The appropriate sparse metadata was added") {
+          REQUIRE(pkg3->SparsePresent("package1::sparse", sparse_ids[2]));
+          REQUIRE(pkg3->SparsePresent("package2::sparse", sparse_ids[3]));
+          REQUIRE(!(pkg3->SparsePresent("package1::sparse", sparse_ids[3])));
+          REQUIRE(!(pkg3->SparsePresent("package2::sparse", sparse_ids[2])));
+        }
+      }
+    }
+    // TODO(JMM): This will simplify once we have dense on block
+    WHEN("We add multiple provides sparse ids to the same package") {
+      for (int i = 0; i < sparse_ids.size(); i++) {
+        pkg1->AddField("sparse", m_sparse_provides[i]);
+      }
+      THEN("We can safely resolve packages") {
+        auto pkg4 = ResolvePackages(packages);
+        AND_THEN("The sparse variable is present") {
+          for (int i = 0; i < sparse_ids.size(); i++) {
+            pkg4->FieldMetadata("sparse", i).SparseEqual(m_sparse_provides[0]);
+          }
+        }
+      }
+    }
+
+    WHEN("We add the same private swarm to two different packages") {
+      pkg1->AddSwarm("swarm", m_private);
+      pkg2->AddSwarm("swarm", m_private);
+      pkg1->AddSwarmValue("value1", "swarm", m_swarmval);
+      pkg2->AddSwarmValue("value2", "swarm", m_swarmval);
+      THEN("We can safely resolve the conflicts") {
+        auto pkg3 = ResolvePackages(packages);
+        AND_THEN("The names are privatley namespaced") {
+          REQUIRE(pkg3->SwarmPresent("package1::swarm"));
+          REQUIRE(pkg3->SwarmPresent("package2::swarm"));
+          REQUIRE(!(pkg3->SwarmPresent("swarm")));
+        }
+        AND_THEN("The swarm values were added appropriately") {
+          REQUIRE(pkg3->SwarmValuePresent("value1", "package1::swarm"));
+          REQUIRE(pkg3->SwarmValuePresent("value2", "package2::swarm"));
+        }
+      }
+    }
+
+    // Do not need to repeat this for sparse/swarm as it is the same
+    // code path
+    WHEN("We add a Requires variable but do not provide it") {
+      pkg1->AddField("dense1", m_requires);
+      pkg2->AddField("dense2", m_requires);
+      THEN("Resolving conflicts raises an error") {
+        REQUIRE_THROWS(ResolvePackages(packages));
+      }
+    }
+    WHEN("We add a dense requires variable and a provides variable") {
+      pkg1->AddField("dense", m_requires);
+      pkg2->AddField("dense", m_provides);
+      THEN("We can safely resolve the conflicts") {
+        auto pkg3 = ResolvePackages(packages);
+        AND_THEN("The provides package is available") {
+          REQUIRE(pkg3->FieldPresent("dense"));
+          REQUIRE(pkg3->FieldMetadata("dense") == m_provides);
+        }
+      }
+    }
+
+    WHEN("We add an overridable variable and nothing else") {
+      pkg1->AddField("dense", m_overridable);
```

Comment:
What is actually happening here?
In principle we already added "dense" with different MetaData above or am I missing where the packages are cleared?

Reply:
`catch2` automatically clears everything between `WHEN`s. Basically `GIVEN` is always called around each new `WHEN` block.

Reply:
That's for the clarification. That's good to know!